docs/dev-process/coding-guidelines.md

```diff
@@ -1,92 +1,172 @@
+
+* Namespace imports should be specified at the top of the file,
+  outside of `namespace` declarations and should be sorted alphabetically.
+
+* Fields should be specified at the top within type declarations.
+
+* File encoding should be `ASCII` (preferred) or `UTF8` (with `BOM`) if absolutely necessary.
+
+### Member Conventions
+
 * Use of `this` is neither encouraged nor discouraged.
-* Avoid more than one blank empty line.
-* Public members should use [doc comments](https://msdn.microsoft.com/en-us/library/b2s063f7.aspx),
-  internal members may use doc comments but it is not encouraged.
+
+* Use `nameof(<member-name>)` instead of `"<member-name>"` whenever possible and relevant.
+
+* Always specify the visibility, even if it's the default (i.e. `private string _foo` not `string _foo`).
+  Visibility should be the first modifier (i.e. `public abstract` not `abstract public`).
+
+* Make members private where possible.
+  Avoid declaring public members unless it's absolutely necessary.
+
 * Public members in a namespace that ends with `Internal`,
   for example `System.Management.Automation.Internal` are not considered a supported public API.
   Such members are necessarily public as implementation details in code shared between C# and PowerShell script,
   or must be available publicly by generated code.
-* File encoding should be ASCII (preferred)
-  or UTF8 (with BOM) if absolutely necessary.
-
-## Preprocessor defines
 
-There are 3 primary preprocessor macros we define during builds:
+### Commenting Conventions
 
-* DEBUG - guard code that should not be included in release builds
-* CORECLR - guard code that differs between Full CLR and CoreCLR
-* UNIX - guard code that is specific to Unix (Linux and macOS)
+* Add comments when changes are not trivial or could be confusing.
 
-Any other preprocessor defines found in the source are used for one-off custom builds,
-typically to help debug specific scenarios.
+* Add comments when a reviewer needs help to understand your changes.
 
-### Runtimes
+* Update existing comments when you are changing the corresponding code.
 
-The PowerShell repo is used to build PowerShell targeting CoreCLR as well as CLR 4.5.
+* Make sure the added/updated comments are accurate and easy to understand.
 
-Code under !CORECLR must build against CLR 4.5.
-We will not accept changes that require a later version of the full CLR.
-In extremely rare cases, we may use reflection to use an API in a later version of the CLR,
-but the feature must robustly handle running with CLR 4.5.
+* Public members must use [doc comments](https://msdn.microsoft.com/en-us/library/b2s063f7.aspx).
+  Internal and private members may use doc comments but it is not required.
 
-We may reject code under !CORECLR without explanation because
-we do not support installation or testing of such code in this repo.
-All new features should support CoreCLR.
-
-## Performance considerations
+## Performance Considerations
```

Comment:
Maybe a note about avoiding exceptions as much as possible? Especially avoid using exceptions for logic (reserve them for unrecoverable errors such as those from the OS or libraries).

Reply:
Good point. Added.